### PR TITLE
Fix #502: Optimize `Array.apply([...])` to `[...]`

### LINKED ISSUE #502: Avoid unnecessary wrapping and array copy when using Array.apply (and potentially List.apply, ...)
Currently, the following code:

``` scala
val x = Array(1,2,3)
```

Is transformed to:

``` scala
val x: Int[] = Array.apply(1, Predef.wrapIntArray([2,3]))
```

Which is as inefficient as it looks. `scalac` has rewriting rules in `Cleanup` to deal with this: https://github.com/scala/scala/blob/d030172d7ef807d85391d32c5456b1b97b15a402/src/compiler/scala/tools/nsc/transform/CleanUp.scala#L536-L545
We could do something similar, keeping in mind the following:
- We have to be careful to not break semantics, see [SI-6611](https://issues.scala-lang.org/browse/SI-6611)
- We'll need more (or better) rewriting rules than `scalac` to deal with all possible cases, for example `scalac` optimizes `Array(1,2,3)` and `Array[String]("foo", "bar")` but does nothing with `Array[Int](1,2,3)`.
- I'd like it if the rewriting was done before `Erasure`, so that it does not need to take into account whatever crazy encoding we'll come up with for arrays of value classes.

The alternative would be to do this and other optimizations like `List(1,2,3)` using macros, see https://groups.google.com/forum/#!topic/scala-internals/5oRYBYBUqMY/discussion and [SI-6247](https://issues.scala-lang.org/browse/SI-6247), is this the direction we want to go in?

### PR DESCRIPTION
**Pull request by nicolasstucki**

### CODE REVIEW

**Comment by nicolasstucki** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/transform/ArrayConstructors.scala [hunk 1]
```diff
@@ -31,17 +33,34 @@ class ArrayConstructors extends MiniPhase {
...
+            Block(t.qualifier :: Nil, expand(targ.head.tpe, dims))
+          case _ => tree
+        }
+      } else {
+        // Optimize `Array.apply([....])` and `Array.apply(..., [....])` into `[...]`
+        tree match {
+          case Apply(Apply(appMeth, Apply(wrapRefArrayMeth, (seqLit: tpd.JavaSeqLiteral) :: Nil) :: Nil), (ct: Apply) :: Nil)
+              if appMeth.symbol.name == nme.apply && appMeth.symbol.owner == defn.ArrayModule &&
+                 defn.WrapArrayMethods().contains(wrapRefArrayMeth.symbol) &&
+                 ct.symbol.maybeOwner.companionModule == defn.ClassTagModule =>
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/ArrayConstructors.scala]: Add a comment explaining why this is the correct thing to check.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/ArrayConstructors.scala]: Modified the code and added comments

File: compiler/src/dotty/tools/dotc/transform/ArrayApply.scala [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+          tpd.JavaSeqLiteral(elem0 :: seqLit.elems, seqLit.elemtpt)
+
+        case _ =>
+          tree
+      }
+
+    } else tree
+  }
+
+  // Only optimize when classtag is `ClassTag.apply` or `ClassTag.{Byte, Boolean, ...}`
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/ArrayApply.scala]: ClassTag.apply is a method that takes a parameter, so I don't see how it's going to end up being checked here, also someone could write `ClassTag.apply({???; classOf[Foo])`, so you can't optimize it away in general.

File: compiler/src/dotty/tools/dotc/transform/ArrayApply.scala [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  private def elideClassTag(ct: Tree)(implicit ctx: Context): Boolean = {
+    ct.symbol.maybeOwner.companionModule == defn.ClassTagModule
+  }
+
+  object CleanTree {
+    def unapply(tree: Tree)(implicit ctx: Context): Some[Tree] = tree match {
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/ArrayApply.scala]: Is this necessary for the common case where the user doesn't manually add type ascriptions or blocks himself ? If it's not necessary then I'd say we shouldn't bother trying to optimize these edge cases.

File: compiler/test/dotty/tools/backend/jvm/ArrayApplyOptTest.scala [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+    test("Array[T]()", List(Op(ICONST_0), IntOp(NEWARRAY, 10), Op(POP), Op(RETURN)))
+  }
+
+  @Test def testArrayGenericApply= {
+    test("""Array("a", "b")""", List(Op(ICONST_2), TypeOp(ANEWARRAY, "java/lang/String"), Op(DUP), Op(ICONST_0), Ldc(LDC, "a"), Op(AASTORE), Op(DUP), Op(ICONST_1), Ldc(LDC, "b"), Op(AASTORE), Op(POP), Op(RETURN)))
+    test("""Array[Object]("a", "b")""", List(Op(ICONST_2), TypeOp(ANEWARRAY, "java/lang/Object"), Op(DUP), Op(ICONST_0), Ldc(LDC, "a"), Op(AASTORE), Op(DUP), Op(ICONST_1), Ldc(LDC, "b"), Op(AASTORE), Op(POP), Op(RETURN)))
+  }
+
+  @Test def testArrayApplyBoolean =
+    test("Array(true, false)", List(Op(ICONST_2), IntOp(NEWARRAY, 4), Op(DUP), Op(ICONST_0), Op(ICONST_1), Op(BASTORE), Op(DUP), Op(ICONST_1), Op(ICONST_0), Op(BASTORE), Op(POP), Op(RETURN)))
```
Comment (smarter) [inline compiler/test/dotty/tools/backend/jvm/ArrayApplyOptTest.scala]: There's a lot of duplication between these tests, that could maybe be avoided with a function `genArrayOpcodes` that takes a few arguments and returns the expected opcodes.

File: compiler/src/dotty/tools/dotc/transform/ArrayApply.scala (hunk 2 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/ArrayApply.scala]: Good point. I will have to inspect the tree to make sure it can be optimized away.

File: compiler/src/dotty/tools/dotc/transform/ArrayApply.scala (hunk 3 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/ArrayApply.scala]: Those are to remove artefacts of inlining. They will be useful if we want to optimize the instantiation an opaque type containing an array (such as in `IArray`).

File: compiler/test/dotty/tools/backend/jvm/ArrayApplyOptTest.scala (hunk 4 above)
Comment (nicolasstucki) [reply compiler/test/dotty/tools/backend/jvm/ArrayApplyOptTest.scala]: Will do

File: compiler/src/dotty/tools/dotc/transform/ArrayApply.scala (hunk 3 above)
Comment (smarter) [reply compiler/src/dotty/tools/dotc/transform/ArrayApply.scala]: I see, but it'd be even nicer if inlining didn't produce such artfacts :)

File: compiler/src/dotty/tools/dotc/transform/ArrayApply.scala (hunk 2 above)
Comment (smarter) [reply compiler/src/dotty/tools/dotc/transform/ArrayApply.scala]: A safe subset is to only optimize if the argument to `ClassTag.apply` is a tree of the form `Literal(Constant(_: Type))` (this requires your phase to be after https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/transform/ClassOf.scala)

File: compiler/src/dotty/tools/dotc/transform/ArrayApply.scala (hunk 3 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/ArrayApply.scala]: Only the `Typed` is an artefact. But now I see it is an artefact of unpacking a varargs into another one. In Scala2 there is an equivalent [cast striped in the same place](https://github.com/scala/scala/blob/d030172d7ef807d85391d32c5456b1b97b15a402/src/compiler/scala/tools/nsc/transform/CleanUp.scala#L540).

File: compiler/src/dotty/tools/dotc/transform/ArrayApply.scala [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+   *  - `ClassTag.apply(classOf[X])`
+   *  - `ClassTag.apply(java.lang.XYZ.Type)`
+   *  - `ClassTag.{Byte, Boolean, ...}`
+   */
+  private def elideClassTag(ct: Tree)(implicit ctx: Context): Boolean = ct match {
+    case Apply(_, rc :: Nil) if ct.symbol == defn.ClassTagModule_apply =>
+      rc match {
+        case _: Literal => true // classOf[X]
+        case rc: RefTree if rc.name == nme.TYPE_ =>
+          val owner = rc.symbol.maybeOwner.companionModule
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/ArrayApply.scala]: You could do `.companionClass` and then check `ScalaBoxedClasses.contains` I think.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/ArrayApply.scala]: Yes, it worked.

File: compiler/src/dotty/tools/dotc/transform/ArrayApply.scala [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+import Symbols._
+import Types._
+import StdNames._
+import ast.Trees._
+import dotty.tools.dotc.ast.tpd
+
+import scala.reflect.ClassTag
+
+
+/** This phase rewrites calls to `Array.apply` to primitive array instantion.
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/ArrayApply.scala]: Typo: instantiation. Also "primitive array" usually means array of primitive so I'd use another term.

Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/transform/ArrayApply.scala]: Updated
